### ISSUE TITLE
[REVIEW]: pymdp: A Python library for active inference in discrete state spaces

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#4098]
**Submitting author:** @conorheins (<a href="http://orcid.org/0000-0002-5884-7728">Conor Heins</a>)
**Repository:** https://github.com/infer-actively/pymdp
**Branch with paper.md** (empty if default branch): master
**Version:** v0.0.5
**Editor:** @emdupre
**Reviewers:** @seankmartin, @patrickmineault
**Archive:** 10.5281/zenodo.6484849

 

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/a6362c8d3e0c953d8ce832664a32f66b/status.svg)](https://joss.theoj.org/papers/a6362c8d3e0c953d8ce832664a32f66b)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/a6362c8d3e0c953d8ce832664a32f66b"><img src="https://joss.theoj.org/papers/a6362c8d3e0c953d8ce832664a32f66b/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/a6362c8d3e0c953d8ce832664a32f66b/status.svg)](https://joss.theoj.org/papers/a6362c8d3e0c953d8ce832664a32f66b)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/infer-actively/pymdp" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@seankmartin & @patrickmineault, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @emdupre know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

## Review checklist for @seankmartin
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/infer-actively/pymdp">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/infer-actively/pymdp">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@conorheins) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

## Review checklist for @patrickmineault
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/infer-actively/pymdp">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/infer-actively/pymdp">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@conorheins) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @seankmartin, @patrickmineault it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.16 s (396.3 files/s, 143004.9 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Python                          30           1995           1963           4615
Jupyter Notebook                11              0           9723           1496
MATLAB                           9            558            790           1406
Markdown                         2             94              0            234
YAML                             2             10             10             42
reStructuredText                 8             29             40             36
DOS Batch                        1              8              1             26
make                             1              4              7              9
-------------------------------------------------------------------------------
SUM:                            64           2698          12534           7864
-------------------------------------------------------------------------------


Statistical information for the repository '0f55aadd4137f41905c03063' was
gathered on 2022/01/25.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
Alexander Kiefer                 2             5              3            0.01
Beren                            4           193             54            0.32
Brennan Klein                    1             1              1            0.00
Conor Heins                     87          8074           2193           13.16
alec-tschantz                   55          8626           6334           19.18
arun                             3            88             73            0.21
conorheins                     237         26192          26172           67.13

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
Alexander Kiefer              5          100.0          6.5                0.00
Beren                        13            6.7         20.0               15.38
alec-tschantz              1245           14.4         18.4               12.21
arun                         13           14.8         11.6                0.00
conorheins                 7297           27.9          7.3                9.09
```

**Comment by whedon** [comment]
PDF failed to compile for issue #4098 with the following error: 
```
 Can't find any papers to compile :-(

```

**Comment by emdupre** [comment]
@whedon generate pdf from branch paper_branch

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch paper_branch. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.04098/joss.04098/10.21105.joss.04098.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.04098/joss.04098/10.21105.joss.04098.pdf) :page_facing_up: :point_left:

**Comment by emdupre** [comment]
@whedon check references from branch paper_branch

**Comment by whedon** [comment]
```
Attempting to check references... from custom branch paper_branch
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1371/journal.pone.0006421 is OK
- 10.1038/s41586-020-2649-2 is OK
- 10.1162/neco_a_01354 is OK
- 10.1080/17588928.2015.1020053 is OK
- 10.1038/s41598-018-38246-3 is OK
- 10.1561/2200000001 is OK
- 10.1167/jov.20.4.12 is OK
- 10.1038/ncomms14218 is OK
- 10.3390/e22050552 is OK
- 10.1162/neco.1992.4.3.415 is OK
- 10.1016/j.neuroimage.2004.07.041 is OK
- 10.1080/01621459.2017.1285773 is OK
- 10.1126/science.1238406 is OK
- 10.1007/s00422-012-0512-8 is OK
- 10.1016/j.tics.2011.11.018 is OK
- 10.1162/neco_a_01357 is OK
- 10.1109/IJCNN48605.2020.9207382 is OK
- 10.1016/0166-2236(83)90190-X is OK
- 10.1523/JNEUROSCI.11-03-00641.1991 is OK
- 10.1162/NECO_a_00912 is OK
- 10.3390/e21030257 is OK
- 10.3389/frobt.2019.00020 is OK
- 10.1007/s00422-018-0785-7 is OK
- 10.1371/journal.pcbi.1007805 is OK
- 10.1109/ICASSP40776.2020.9054364 is OK
- 10.1016/B978-0-12-372560-8.X5000-1 is OK
- 10.1038/nrn2787 is OK
- 10.1016/j.jmp.2020.102447 is OK
- 10.1016/S0004-3702(98)00023-X is OK
- 10.1162/neco_a_01351 is OK
- 10.1093/cercor/bhu159 is OK
- 10.1016/j.drugalcdep.2020.108208 is OK
- 10.1503/jpn.200032 is OK
- 10.31234/osf.io/b4jm6 is OK
- 10.1016/j.neuroimage.2008.02.054 is OK
- 10.1016/j.jmp.2020.102348 is OK
- 10.1007/978-3-030-64919-7_1 is OK
- 10.1016/j.neubiorev.2016.06.022 is OK
- 10.1162/neco.1995.7.5.889 is OK
- 10.1037/13536-000 is OK
- 10.1016/j.neunet.2021.05.010 is OK
- 10.1109/LRA.2021.3090015 is OK
- 10.1016/B978-0-323-88506-5.50066-8 is OK
- 10.1007/s00422-019-00805-w is OK
- 10.3390/e23020198 is OK
- 10.1080/10407413.2021.1965480 is OK
- 10.1098/rsif.2013.0475 is OK
- 10.7554/eLife.41703 is OK
- 10.1162/neco_a_00999 is OK

MISSING DOIs

- 10.1162/neco_a_01427 may be a valid DOI for title: Chance-Constrained Active Inference
- 10.1017/s0140525x19001353 may be a valid DOI for title: Generative models as parsimonious descriptions of sensorimotor loops
- 10.7551/mitpress/7111.003.0009 may be a valid DOI for title: Perceptions as hypotheses
- 10.31234/osf.io/2rzu5 may be a valid DOI for title: Active inference, selective attention, and the cocktail party problem
- 10.1016/j.schres.2021.07.032 may be a valid DOI for title: Everything is connected: Inference and attractors in delusions
- 10.31234/osf.io/bsuh8 may be a valid DOI for title: Long-term stability of computational parameters during approach-avoidance conflict in a transdiagnostic psychiatric patient sample

INVALID DOIs

- None
```

**Comment by emdupre** [comment]
Hi everyone :wave: Thanks again for agreeing to review this submission ! The review will take place in this issue.

Whenever possible, please open relevant issues on the [linked software repository ](https://github.com/infer-actively/pymdp)(and cross-link them with this issue) rather than discussing them here. This helps to make sure that feedback is translated into actionable items to improve the software. 

If you aren't sure how to get started, please see the [Reviewing for JOSS guide](https://joss.readthedocs.io/en/latest/reviewer_guidelines.html) -- and, of course, feel free to ping me with any questions !

**Comment by seankmartin** [comment]
Cool, thanks!
I should be able to get to doing the review fairly soon

**Comment by seankmartin** [comment]
@whedon commands

**Comment by whedon** [comment]
Here are some things you can ask me to do:

```
# List Whedon's capabilities
@whedon commands

# List of editor GitHub usernames
@whedon list editors

# List of reviewers together with programming language preferences and domain expertise
@whedon list reviewers

EDITORIAL TASKS

# Compile the paper
@whedon generate pdf

# Compile the paper from alternative branch
@whedon generate pdf from branch custom-branch-name

# Ask Whedon to check the references for missing DOIs
@whedon check references

# Ask Whedon to check repository statistics for the submitted software
@whedon check repository

```

**Comment by seankmartin** [comment]
I'm in progress with the review.
I've raised a non-blocking issue about some missing API documentation here https://github.com/infer-actively/pymdp/issues/67.
The paper is missing an explicit summary section, which I think should be added.
Will post a final update after checking functionality!

**Comment by seankmartin** [comment]
@whedon remind @seankmartin in 4 days

**Comment by whedon** [comment]
I'm sorry @seankmartin, I'm afraid I can't do that. That's something only editors are allowed to do.

**Comment by patrickmineault** [comment]
I'm going to wait until @seankmartin completes a first pass until I look at this.

**Comment by emdupre** [comment]
@whedon remind @seankmartin in 3 days

**Comment by whedon** [comment]
Reminder set for @seankmartin in 3 days

**Comment by emdupre** [comment]
@seankmartin I've set the reminder :) Thanks for starting on this and cross-linking the opened issue !

**Comment by whedon** [comment]
:wave: @seankmartin, please update us on how your review is going (this is an automated reminder).

**Comment by whedon** [comment]
:wave: @seankmartin, please update us on how your review is going (this is an automated reminder).

**Comment by whedon** [comment]
:wave: @patrickmineault, please update us on how your review is going (this is an automated reminder).

**Comment by conorheins** [comment]
> I'm in progress with the review. I've raised a non-blocking issue about some missing API documentation here [infer-actively/pymdp#67](https://github.com/infer-actively/pymdp/issues/67). The paper is missing an explicit summary section, which I think should be added. Will post a final update after checking functionality!

Hi @emdupre, I'm wondering if we can already start responding to the issues raised by @seankmartin ? Or should we wait until both reviewers are finished before we start responding? Thanks :)

**Comment by emdupre** [comment]
Hi @conorheins, thanks for confirming on this ! It's entirely up to you whether you'd like to address @seankmartin and @patrickmineault's reviews as they come in or all at once. 

After the reviews are in, I'll check-in with you on the status of your response. If you've been addressing comments as they come in, this will usually mean confirming that you're ready to receive additional reviewer feedback on your revisions. If instead you've elected to wait until all comments are available, then I'll ask for your anticipated timeline for making revisions.

If you have any other questions, please don't hesitate to ask.

**Comment by seankmartin** [comment]
Ok I'm finished a first pass now.
Sincere apologies for the delay in particular to @patrickmineault since you were waiting for me.

I am very happy with the functionality of the package. The tests all pass and are useful, and I have manually verified some outputs in places which also were correct.
I have also run all of the demos/tutorials. I think the cue chaining (Epistemic Chaining) demo is particularly well written and is probably the best branching point for doing your own stuff with pymdp.
I like that the env and agent classes follow a similar pattern to OpenAI Gym, just makes it easier to work with.

My main concerns would be as before then (about API docs and no explicit summary section in paper). To reiterate slightly:
The package in general is very well documented, such as in the `control.py` file where the docstrings are very informative and have API docs on your read the docs website.
I think that the Agent and Env class are critical and that they should have class docstrings and should be included on the read the docs website

Overall I think the package is great and clearly has had an lot of effort put into it!

**Comment by conorheins** [comment]
Hi @seankmartin , thanks for your comments and reviewing our package. We believe we have addressed the main API issues you raised. The latest version of the RTD page with new sections for `Agent` and `Env` API documentation can be found here: https://pymdp-rtd.readthedocs.io/en/paper_branch.

I haven't addressed your issue about the README.md badges yet: https://github.com/infer-actively/pymdp/issues/68 but I thought I'd just post an update here now, to signal that we believe we've addressed Issue https://github.com/infer-actively/pymdp/issues/67 and the main brunt of your review.

Note we've also added an explicit summary section in the paper, also as requested.

**Comment by seankmartin** [comment]
Thanks for the update @conorheins - the docs look good!

No worries about the README badges, that is just a suggestion, not required by any means.

**Comment by emdupre** [comment]
:wave: Hi @patrickmineault, please let us know how your review is going !

**Comment by emdupre** [comment]
@editorialbot set paper_branch as branch

**Comment by emdupre** [comment]
@editorialbot generate pdf

**Comment by patrickmineault** [comment]
Apologies, will get to it this weekend.

**Comment by conorheins** [comment]
Hi @patrickmineault, just checking in on this review

**Comment by patrickmineault** [comment]
Apologies for the lateness of my review - this is great! The authors very clearly explain the need in the companion paper - currently, the mysteries of the free energy principle are accessible to few and tucked away in a corner of the Matlab package  SPM. By making their methods widely accessible across a wide audience, they're opening the field. I really appreciate the effort that went into making the API compatible with OpenAI gym - and would love to see deeper comparisons of RL and POMDP free energy agents. As a side note, the star history (+150% stars since the start of the year) makes it clear that people already find it useful.

The docs are generally great, and the emphasis on tutorials should make it easy for anyone to get up and running. I've opened a few very minor, and optional issues on the Github repo about the first two tutorials. I see that you've addressed @seankmartin 's issues with Agent being undocumented in the latest RDC, which was also a point of contention.

It's ready to ship as far as I'm concerned.

**Comment by conorheins** [comment]
Hi @patrickmineault,

Thanks very much for the kinds words/ thoughtful reviews, particularly the attention you gave to the tutorials - those were all fantastic suggestions and I will get down to incorporating your changes this weekend at the latest.

**Comment by emdupre** [comment]
Thank you @seankmartin and @patrickmineault for your thoughtful reviews !

@conorheins, I see that https://github.com/infer-actively/pymdp/issues/67, https://github.com/infer-actively/pymdp/issues/68, https://github.com/infer-actively/pymdp/issues/71, and https://github.com/infer-actively/pymdp/issues/72 are all open from this review process, though the reviewers are quite consistent in noting that these are optional asks. 

Could you please confirm when you're finished incorporating this reviewer feedback and are ready to move forward with the submission ? I'll also provide a few editorial comments shortly.

**Comment by emdupre** [comment]
@editorialbot check references

**Comment by conorheins** [comment]
Hi @emdupre,

Even though they're optional, I can address those remaining issues this weekend (namely https://github.com/infer-actively/pymdp/issues/71, and https://github.com/infer-actively/pymdp/issues/72) before we fully proceed -- I'd also like to edit an affiliations and a funding acknowledgement, if that's okay. As soon as I'm finished I'll post here to confirm that I'm all set.

And a quick question regarding the closing of issues -- I was planning to wait until paper_branch was merged with master before I closed https://github.com/infer-actively/pymdp/issues/67 (and the others of course, but that's the only one I've already addressed) but I can already close it now if that's preferable for the review process. 

Thanks,
Conor

**Comment by emdupre** [comment]
> As soon as I'm finished I'll post here to confirm that I'm all set.

That sounds perfect ! Thank you.

> I was planning to wait until paper_branch was merged with master before I closed https://github.com/infer-actively/pymdp/issues/67 (and the others of course, but that's the only one I've already addressed) but I can already close it now if that's preferable for the review process. Would that be preferable?

It makes it a bit clearer that you consider the issue(s) entirely address, though closing before merging in the branch is not a requirement -- especially since these are optional additions to the project !

I also have a few requests on the paper itself. If you could please:
- [ ] add in DOIs to those references noted with "DOI missing" in the most [recent editorialbot comment](https://github.com/openjournals/joss-reviews/issues/4098#issuecomment-1092161263). The bot has suggested a potential DOI for each reference, but please confirm if these are correct !
- [ ] update the ForneyLab citation to match [their citation.bib file](https://github.com/biaslab/ForneyLab.jl/blob/master/CITATION.bib) ?
- [ ] standardize the formatting of arXiv references; e.g., Baioumy et al (2021) is cited with `arXiv. <arXiv-URL>` while Laar et al (2021) is cited with `arXiv Preprint <arXiv:URI>` 

More generally, too: Could you clarify the relationship between this paper and [the companion arXiv preprint](https://arxiv.org/abs/2201.03904) ? There are partially overlapping sections, and I want to make sure I understand the relationship between these two resources !

**Comment by conorheins** [comment]
Hi @emdupre,

Thanks for pointing out these issues, I've attempted to address them all. Let me know if there are any further issues. For the missing DOIs, some of the suggestions were correct, some of them were not, so I just manually added them all in. I hope I did so correctly :) 

- [x] add in DOIs to those references noted with "DOI missing" in the most https://github.com/openjournals/joss-reviews/issues/4098#issuecomment-1092161263. The bot has suggested a potential DOI for each reference, but please confirm if these are correct 

- [x] update the ForneyLab citation to match [their citation.bib file](https://github.com/biaslab/ForneyLab.jl/blob/master/CITATION.bib) ?

- [x] standardize the formatting of arXiv references; e.g., Baioumy et al (2021) is cited with arXiv. <arXiv-URL> while Laar et al (2021) is cited with arXiv Preprint <arXiv:URI>

And finally, regarding the clarification between preprint companion and this JOSS submission -- we wrote the longer, technical version first, to be out there for those who are interested in the mathematical background of these models  (POMDPs) that one can build in `pymdp`. Then we decided to make the submission for JOSS  just a shortened, more cleaned up / tighter version of the arXiv version, that focuses on the software itself and its goals/advantages. That's why you'll notice groups of overlapping text (particularly in the earlier sections), because for all intents and purposes the arXiv preprint version is just a longer, more technical version of the JOSS submission. 

So given all that: would it be okay for us to keep the arXiv preprint online, as the more technical version of the JOSS article, but then ask users to officially cite the JOSS version? I think it would be helpful to have that longer, more mathematical version with the appendices available as a pedagogical resource. But obviously we would want the JOSS version to be the "official" version that is cited.

**Comment by conorheins** [comment]
> Hi @emdupre,
> 
> Even though they're optional, I can address those remaining issues this weekend (namely [infer-actively/pymdp#71](https://github.com/infer-actively/pymdp/issues/71), and [infer-actively/pymdp#72](https://github.com/infer-actively/pymdp/issues/72)) before we fully proceed -- I'd also like to edit an affiliations and a funding acknowledgement, if that's okay. As soon as I'm finished I'll post here to confirm that I'm all set.
> 
> And a quick question regarding the closing of issues -- I was planning to wait until paper_branch was merged with master before I closed [infer-actively/pymdp#67](https://github.com/infer-actively/pymdp/issues/67) (and the others of course, but that's the only one I've already addressed) but I can already close it now if that's preferable for the review process. Would that be preferable?
> 
> Thanks, Conor

Oh and I forgot to confirm the resolution of the issues. Regarding the optional requests raised by the reviewers, I've now addressed nearly all of them, and just left some minor bits open that I'd like to work out in the future. But otherwise I am ready to proceed with the submission.

**Comment by emdupre** [comment]
@editorialbot generate pdf

**Comment by emdupre** [comment]
@editorialbot check references

**Comment by emdupre** [comment]
Thank you, @conorheins !

I'm confirming the best course of action under JOSS policy for the arXiv companion piece, and I will have an answer for you -- along with next steps -- shortly.

**Comment by emdupre** [comment]
Hi @conorheins ,

Thank you for your patience on this ! I've consulted with other JOSS editors and have outlined a list of next steps that I hope will be helpful. I'm detailing them below, and please do let me know if you have any questions in addressing them.

- [ ] It would be best to directly cite the arXiv companion piece in the JOSS article such that the relationship between the two is clearly documented in the publication. This will help readers to consult the appropriate resource, as necessary, and maintain a record of both pieces. I would suggest including this citation in the `Statement of Need` section to expand on your stated goal of increasing the "accessibility and exposure of the active inference framework to researchers, engineers, and developers with diverse disciplinary backgrounds."
- [ ] Thank you for standardizing the arXiv references ! I received a bit more information on the best practice in formatting these. Could you please modify them to include the `eprint`, `archivePrefix`, and `primaryClass` fields? So, for example, for [millidge2021understanding](https://github.com/infer-actively/pymdp/blob/9da57b57683bae1be65db8145b0c50c4862388ce/paper/paper.bib#L61-L66) the reference would be:
  ```
  @misc{millidge2021understanding,
  title={Understanding the origin of information-seeking exploration in probabilistic objectives for control},
  author={Millidge, Beren and Tschantz, Alexander and Seth, Anil and Buckley, Christopher},
  eprint={2103.06859v7},
  archivePrefix={arXiv},
  primaryClass={cs.LG}, 
  year={2021}
  }
    ```
 - [ ] For [`van2021chance`](https://github.com/infer-actively/pymdp/blob/9da57b57683bae1be65db8145b0c50c4862388ce/paper/paper.bib#L30-L35), could you update the reference from the arXiv preprint to the published _Neural Computation_ article:
   ```
   @article{10.1162/neco_a_01427,
    author={van de Laar, Thijs and Senoz, Ismail and {\"O}z{\c{c}}elikkale, Ay{\c{c}}a and Wymeersch, Henk},
    title = {Chance-Constrained Active Inference},
    journal = {Neural Computation},
    volume = {33},
    number = {10},
    pages = {2710-2735},
    year = {2021},
    month = {09},
    issn = {0899-7667},
    doi = {10.1162/neco_a_01427},
    url = {https://doi.org/10.1162/neco\_a\_01427},
    eprint = {https://direct.mit.edu/neco/article-pdf/33/10/2710/1963501/neco\_a\_01427.pdf},
    }
   ``` 
 
Pending these changes, could you then please:

- [ ] Make a tagged release of your software, and list the version tag of the archived version here.
- [ ] Archive the reviewed software in Zenodo or a similar service (e.g., figshare, an institutional repository)
- [ ] Check the archival deposit (e.g., in Zenodo) has the correct metadata (e.g., the JOSS paper and software archive have the same title and author listing)
- [ ] List the DOI of the archived version here.

I can then move forward with accepting the submission 🚀

**Comment by conorheins** [comment]
Hi @emdupre , I've now finished the re-formatting steps and changes to Statement of Need that you suggested. 

Before I move onto the second set of steps (e.g. archiving the reviewed software in Zenodo, ... .), I just had a clarifying question about the order in which I do these steps.

The first step is "Make a tagged release of your software, and list the version tag of the archived version here".

Does this "archived version" refer to the archived version that will be generated in the subsequent steps (with Zenodo, etc.), or something else that I'm not understanding? This is probably obvious for someone who has worked with Zenodo or similar services but this is new for me, so before proceeding I just want to know what exactly I should be listing in the version tag of the tagged release.

**Comment by emdupre** [comment]
@editorialbot generate pdf

**Comment by emdupre** [comment]
Hi @conorheins,

Thank you for addressing these points ! There's a typo [here](https://github.com/infer-actively/pymdp/blob/137cec4fff77ea29ac9a922351d7671d79299823/paper/paper.bib#L68) that's preventing PDF compilation. If you can remove the extra `}` character, we should be able to re-generate the PDF !

> Does this "archived version" refer to the archived version that will be generated in the subsequent steps (with Zenodo, etc.), or something else that I'm not understanding? This is probably obvious for someone who has worked with Zenodo or similar services but this is new for me, so before proceeding I just want to know what exactly I should be listing in the version tag of the tagged release.

This is a good question; thank you for confirming ! What I mean is that when you create a new release of your software, this release is the one that is then archived in subsequent steps (e.g. on Zenodo using [this workflow](https://docs.github.com/en/repositories/archiving-a-github-repository/referencing-and-citing-content)), as you suggested. The motivation is that you have [a tagged release](https://docs.github.com/en/repositories/releasing-projects-on-github/managing-releases-in-a-repository) that includes the state of the software as it is described in the JOSS paper, and by archiving this tagged release it will have a persistent identifier; e.g. a DOI, that we can include in the JOSS publication.

Hopefully that helps to answer your questions, but please let me know if there's anything else I can clarify !

**Comment by conorheins** [comment]
Hi @emdupre, thanks for the clarification. 

As requested, here is a link to the latest release: [v0.0.5](https://github.com/infer-actively/pymdp/tree/v0.0.5)

And here is the DOI (from Zenodo): 10.5281/zenodo.6484849

I have also verified that the Zenodo archive version has the correct metadata (title and author list).

**Comment by emdupre** [comment]
@editorialbot  generate pdf

**Comment by emdupre** [comment]
@editorialbot generate pdf

**Comment by emdupre** [comment]
@editorialbot set version v0.0.5

**Comment by emdupre** [comment]
@editorialbot set v0.0.5 as version

**Comment by emdupre** [comment]
@editorialbot check references

**Comment by emdupre** [comment]
@editorialbot set 10.5281/zenodo.6484849 as archive

**Comment by emdupre** [comment]
@editorialbot check references

**Comment by emdupre** [comment]
Thank you, @conorheins !

A few remaining points:
- [ ] The [software archive](https://doi.org/10.5281/zenodo.6484849) title is currently `infer-actively/pymdp: 0.0.5` -- could you please update it to `pymdp: A Python library for active inference in discrete
state spaces` to match the JOSS paper ?
- [ ] Similarly, the [software archive](https://doi.org/10.5281/zenodo.6484849) author listing is ` conorheins; Alexander Tschantz; Arun-Niranjan; Beren; mahault` when I would expect it to be `Conor Heins, Beren Millidge, Daphne Demekas, Brennan Klein, Karl Friston, Iain D. Couzin, and Alexander Tschantz`. Could you please update the author formatting to match the the formatting included in the paper (e.g., changing `conorheins` to `Conor Heins`) and confirm that you intended to have a different author list between the software paper and archive ?
- [ ] For the references that editorialbot identifies as "missing," I believe that you have included the `baltieri` reference twice in your bib list, with the second `baltieri2019pid` reference corresponding to the later version of the same article. `baioumy2021towards` also has a published version, but I assume you intended to keep the arXiv version listed s.t. readers can access the non-paywalled version. Please let me know if either of these assessments are incorrect.

These first two edits can be made directly on Zenodo by editing the metadata associated with the archive, as [described here](https://help.zenodo.org/#technical) !

**Comment by conorheins** [comment]
Hi @emdupre,

Re: software archive metadata. Sorry about that, that's strange. I changed those metadata (title / author list) to the correct versions after I made the Zenodo release, but it seems it doesn't show up for the public-facing version, presumably because I didn't click "Publish" yet, only "Save". I've now clicked "Publish" so all the metadata should be consistent now. Let me know if you still notice any discrepancies. The title and author list should be identical to those on the paper now, hopefully.

re: Baltieri referrences -- `baltieri2019pid` (an entropy paper) is actually a different paper than `baltieri2019` (a commentary on another paper), so I think that should be fine.

re: `baouimy2021towards` -- you are correct that I intended to keep the arXiv version listed.

**Comment by emdupre** [comment]
Great, thank you ! Only the `@baltieri2019pid` citation is currently included in the paper.md, so I wanted to confirm their relationship. I think, though, that having an uncited reference in the .bib file is not against JOSS policy, so we can move forward as-is.

I'm now happy to recommend pymdp to the EIC team for publication, and I just want to add my congratulations to you on such an impressive effort !

**Comment by emdupre** [comment]
@editorialbot recommend-accept

**Comment by conorheins** [comment]
Hi @emdupre,

Awesome, thank you! This is great news.

Before the paper is published, would now be a good time for me to just pass it around the co-authors to make sure everyone's name / affiliations are listed properly? Or is there an equivalent of a 'final proofs' stage coming later, at which small edits like that would be appropriate?

Thanks,
Conor

**Comment by emdupre** [comment]
Please pass it around now ! This is effectively the "final proofs" stage, so we can still incorporate these corrections as well as any others that the EiCs suggest.

Please let us know when this is confirmed on your side !

**Comment by conorheins** [comment]
Ok sounds good, will do!

**Comment by emdupre** [comment]
When you're updating affiliations, @conorheins, can you please also:

 - [ ] Remove the uncited `baltieri2019generative` reference from the .bib file
 - [ ] Update the `baouimy2021towards` reference to the paywalled version, doi: 10.1007/978-3-030-93736-2_48

I received additional guidance that we will need the the version-of-record in the JOSS publication.

**Comment by conorheins** [comment]
Hi @emdupre, I've gotten confirmation that affiliations/names are all correct :) I've also updated .bib file as requested

**Comment by emdupre** [comment]
@editorialbot check references

**Comment by emdupre** [comment]
@editorialbot set master as branch

**Comment by emdupre** [comment]
@editorialbot check references

**Comment by emdupre** [comment]
Thank you, @conorheins ! I have one last reference request, since the included Hipólito reference was published in _Frontiers_ over the weekend. Can you please update [this reference](https://github.com/infer-actively/pymdp/blob/d2b1d2b5cae4b9af9255d821a30f39e4be8dfd89/paper/paper.bib#L646-L651) to:

```
@article{hipolito2021enactive,
author = {Hip{\'o}lito, In{\^e}s and van Es, Thomas},
title = {Enactive-Dynamic Social Cognition and Active Inference},
journal = {Frontiers in Psychology},
volume = {13},
year = {2022},
url = {https://www.frontiersin.org/article/10.3389/fpsyg.2022.855074},
doi = {10.3389/fpsyg.2022.855074},
issn = {1664-1078}
}
```

**Comment by conorheins** [comment]
Hi @emdupre , I removed that reference as well as several others that weren't used here (some of them were left over from the arxiv version). Thanks

**Comment by emdupre** [comment]
@editorialbot check references

**Comment by emdupre** [comment]
@openjournals/joss-eics, this is ready for final review !

**Comment by kyleniemeyer** [comment]
@editorialbot recommend-accept

**Comment by kyleniemeyer** [comment]
Hi @conorheins, I'm the AEIC on duty this week doing some final checks.

In your article, can you remove the explicit arXiv URL from the end of the statement of need section? The citation you included is sufficient and includes the URL.

Similarly, the article already links to the software repository, so I think you should remove the URL pointing to that in the same paragraph. If you really want to keep a link in the text, perhaps just make the word "repository" a hyperlink like `[repository](https://github.com/infer-actively/pymdp)`.

In the Introduction first paragraph, can you wrap "e.g." with commas, like "... in, e.g., psychophysics ...". Similarly, please add a comma after "e.g." in later uses.

It looks like the Parr et al. 2020 article from _Cerebral Cortex_ is missing a DOI, can you check if one is available?

**Comment by conorheins** [comment]
Hi @kyleniemeyer, thanks for the final checks. I have made the changes you requested, let me know if anything else needs fixing.

**Comment by kyleniemeyer** [comment]
@editorialbot generate pdf

**Comment by kyleniemeyer** [comment]
@editorialbot accept

**Comment by kyleniemeyer** [comment]
Congratulations @conorheins on your article's publication in JOSS! ([Conor Heins](http://orcid.org/0000-0002-5884-7728))

Many thanks to @seankmartin and @patrickmineault for reviewing this, and @emdupre for editing.

**Comment by conorheins** [comment]
Thanks very much (on behalf of all the authors) for the great news, and thank you again @seankmartin and @patrickmineault for your helpful reviews. And finally many thanks @emdupre for all your help/patience during the review process!
